File: core/src/main/java/ru/pocketbyte/locolaser/resource/PlatformSetResources.java
package ru.pocketbyte.locolaser.resource;

import ru.pocketbyte.locolaser.config.WritingConfig;
import ru.pocketbyte.locolaser.resource.entity.ResMap;
import ru.pocketbyte.locolaser.summary.FileSummary;

import java.io.IOException;
import java.util.Set;

public class PlatformSetResources implements PlatformResources {

    private Set<PlatformResources> mResources;

    public PlatformSetResources(Set<PlatformResources> resources) {
        mResources = resources;
    }

    @Override
    public ResMap read(Set<String> locales) {
        ResMap map = new ResMap();

        for (PlatformResources resource: mResources)
            map.merge(resource.read(locales));

        return map;
    }

    @Override
    public void write(ResMap map, WritingConfig writingConfig) throws IOException {
        for (PlatformResources resource: mResources)
            resource.write(map, writingConfig);
    }

    @Override
    public FileSummary summaryForLocale(String locale) {
        long bytes = 0;
        StringBuilder hash = new StringBuilder();

        for (PlatformResources resource: mResources) {
            FileSummary summary = resource.summaryForLocale(locale);
            bytes += summary.bytes;
            hash.append(summary.hash);
        }

        return new FileSummary(bytes, hash.toString());
    }
}

File: core/src/main/java/ru/pocketbyte/locolaser/resource/PlatformSetResources.kt
package ru.pocketbyte.locolaser.resource

import ru.pocketbyte.locolaser.config.ExtraParams
import ru.pocketbyte.locolaser.resource.entity.ResMap
import ru.pocketbyte.locolaser.resource.entity.merge
import ru.pocketbyte.locolaser.resource.formatting.FormattingType
import ru.pocketbyte.locolaser.resource.formatting.MixedFormattingType
import ru.pocketbyte.locolaser.resource.formatting.NoFormattingType
import ru.pocketbyte.locolaser.summary.FileSummary

import java.io.IOException

class PlatformSetResources(
        private val resources: Set<PlatformResources>
) : PlatformResources {

    override val formattingType: FormattingType
        get() {
            val firstItemType = if (resources.isEmpty()) {
                NoFormattingType
            } else {
                resources.elementAt(0).formattingType
            }

            return if (resources.find { it.formattingType != firstItemType } != null) {
                MixedFormattingType
            } else {
                firstItemType
            }
        }

    override fun read(locales: Set<String>?, extraParams: ExtraParams?): ResMap? {
        return ResMap().apply {
            resources.forEach {
                merge(it.read(locales, extraParams))
            }
        }
    }

    @Throws(IOException::class)
    override fun write(resMap: ResMap, extraParams: ExtraParams?) {
        resources.forEach {
            it.write(resMap, extraParams)
        }
    }

    override fun summaryForLocale(locale: String): FileSummary {
        var bytes: Long = 0
        val hash = StringBuilder()

        resources.map { it.summaryForLocale(locale) }.forEach {
            bytes += it.bytes
            hash.append(it.hash)
        }

        return FileSummary(bytes, hash.toString())
    }
}
